pkg/controller/garbagecollector/metaonly/metaonly.go

```diff
@@ -0,0 +1,130 @@
+/*
+Copyright 2015 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metaonly
+
+import (
+	gojson "encoding/json"
+	"fmt"
+	"io"
+	"reflect"
+	"strings"
+
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/api/v1"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/util/json"
+)
+
+// MetaOnly allows decoding only the apiVersion, kind, and metadata fields of
+// JSON data.
+// TODO: enable meta-only decoding for protobuf.
+type MetaOnly struct {
+	unversioned.TypeMeta `json:",inline"`
+	v1.ObjectMeta        `json:"metadata,omitempty"`
```

Comment:
We don't have a mechanism to convert TPR or unrecognized types anyway.
Which means that swagger and/or api discovery has to communicate that info
if we do have to solve that.  By putting this in we probably aren't making
that problem worse... except for our own types.

On Mon, Jul 25, 2016 at 11:20 PM, Chao Xu <notifications@github.com> wrote:

> In pkg/controller/garbagecollector/metaonly/metaonly.go
> <https://github.com/kubernetes/kubernetes/pull/28480#discussion_r72182360>
> :
>
> > +	"io"
> > +	"reflect"
> > +	"strings"
> > +
> > +	"k8s.io/kubernetes/pkg/api/unversioned"
> > +	"k8s.io/kubernetes/pkg/api/v1"
> > +	"k8s.io/kubernetes/pkg/runtime"
> > +	"k8s.io/kubernetes/pkg/util/json"
> > +)
> > +
> > +// MetaOnly allows decoding only the apiVersion, kind, and metadata fields of
> > +// JSON data.
> > +// TODO: enable meta-only decoding for protobuf.
> > +type MetaOnly struct {
> > +	unversioned.TypeMeta `json:",inline"`
> > +	v1.ObjectMeta        `json:"metadata,omitempty"`
>
> I think the danger is that MetaOnly won't work if we use incompatible
> ObjectMeta across different groups in the future. @smarterclayton
> <https://github.com/smarterclayton> @lavalamp
> <https://github.com/lavalamp> is this going to happen in the near future?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/28480/files/cca957905d7ed2af91d0444ecb3dd717ac8c9825#r72182360>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p3ws2Zajl7FyzwA3YwRNuGEp5X4iks5qZXzpgaJpZM4JEvyT>
> .
>